ops/model.py

```diff
@@ -1577,7 +1609,10 @@ def action_get(self):
         return self._run('action-get', return_output=True, use_json=True)
 
     def action_set(self, results):
-        self._run('action-set', *["{}={}".format(k, v) for k, v in results.items()])
+        # The Juju action-set hook tool cannot interpret nested dicts, the _flatten helper here is
+        # and convenience method to flatten out any nested dict structures into a dotted notation.
+        flat_results = _flatten(results)
+        self._run('action-set', *["{}={}".format(k, v) for k, v in flat_results.items()])
```

Comment:
Do we want to do any validation of the keys and values in the dict before flattening them.

E.g., do we want to handle a case like ...

```{"foo": {"bar": "I have a long string. Which includes punctuation. Which is not okay here."}```

Reply:
Should be easy enough to do that from within the `_flatten` method, and raise a `ValueError`. Which characters are *not* permitted in the value?

Reply:
Added some validation of keys in a later commit.